File: Problems 3400-3499/3443. Maximum Manhattan Distance After K Changes/solution.py
class Solution:
    def maxDistance(self, s: str, k: int) -> int:
        latitude, longitude = 0, 0
        res = 0
        
        for i in range(len(s)):
            if s[i] == 'N':
                longitude += 1
            elif s[i] == 'S':
                longitude -= 1
            elif s[i] == 'W':
                latitude += 1
            else:
                latitude -= 1
            res = max(res, min(abs(longitude) + abs(latitude) + 2 * k, i + 1))

        return res

        # res = 0
        # for diag in ["NW", "NE", "SW", "SE"]:
        #     dist = 0
        #     kLeft = k
        #     for c in s:
        #         if c in diag or kLeft > 0:
        #             dist += 1
        #             kLeft -= c not in diag
        #         else:
        #             dist -= 1
        #         res = max(res, dist)
        # return res

File: Problems 3400-3499/3443. Maximum Manhattan Distance After K Changes/solution.ts
function maxDistance(s: string, k: number): number {
    // [N, S] cancel each other out, [E, W] cancel each other out
    // you can treat these pairs independently
    // Modify the least frequent direction from each pair
    let lat = 0;
    let long = 0;
    let res = 0;
    for (let i = 0; i < s.length; i++) {
        const dir = s[i];
        if (dir === 'N') {
            long++;
        } else if (dir === 'S') {
            long--;
        } else if (dir === 'E') {
            lat++;
        } else {
            lat--;
        }
        res = Math.max(res, Math.min(Math.abs(lat) + Math.abs(long) + 2*k, i + 1));
    }

    return res;
};